modules/dnn/src/net_impl_backend.cpp

```diff
@@ -0,0 +1,200 @@
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+
+#include "net_impl.hpp"
+#include "legacy_backend.hpp"
+
+namespace cv {
+namespace dnn {
+CV__DNN_INLINE_NS_BEGIN
+
+
+Ptr<BackendWrapper> Net::Impl::wrap(Mat& host)
+{
+    if (preferableBackend == DNN_BACKEND_OPENCV && preferableTarget == DNN_TARGET_CPU)
+        return Ptr<BackendWrapper>();
+
+    MatShape shape(host.dims);
+    for (int i = 0; i < host.dims; ++i)
+        shape[i] = host.size[i];
+
+    void* data = host.data;
+    if (backendWrappers.find(data) != backendWrappers.end())
+    {
+        Ptr<BackendWrapper> baseBuffer = backendWrappers[data];
+        if (preferableBackend == DNN_BACKEND_OPENCV)
+        {
+#ifdef HAVE_OPENCL
+            CV_Assert(IS_DNN_OPENCL_TARGET(preferableTarget));
+            return OpenCLBackendWrapper::create(baseBuffer, host);
+#else
+            CV_Error(Error::StsInternal, "");
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_HALIDE)
+        {
+            CV_Assert(haveHalide());
+#ifdef HAVE_HALIDE
+            return Ptr<BackendWrapper>(new HalideBackendWrapper(baseBuffer, shape));
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019)
+        {
+            CV_ERROR_DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019;
+        }
+        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+        {
+            return wrapMat(preferableBackend, preferableTarget, host);
+        }
+        else if (preferableBackend == DNN_BACKEND_WEBNN)
+        {
+#ifdef HAVE_WEBNN
+            return wrapMat(preferableBackend, preferableTarget, host);
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_VKCOM)
+        {
+#ifdef HAVE_VULKAN
+            return Ptr<BackendWrapper>(new VkComBackendWrapper(baseBuffer, host));
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_CUDA)
+        {
+            CV_Assert(haveCUDA());
+#ifdef HAVE_CUDA
+            switch (preferableTarget)
+            {
+            case DNN_TARGET_CUDA:
+                return CUDABackendWrapperFP32::create(baseBuffer, shape);
+            case DNN_TARGET_CUDA_FP16:
+                return CUDABackendWrapperFP16::create(baseBuffer, shape);
+            default:
+                CV_Assert(IS_DNN_CUDA_TARGET(preferableTarget));
+            }
+#endif
+        }
+        else
+            CV_Error(Error::StsNotImplemented, "Unknown backend identifier");
+    }
+
+    Ptr<BackendWrapper> wrapper = wrapMat(preferableBackend, preferableTarget, host);
+    backendWrappers[data] = wrapper;
+    return wrapper;
+}
+
+
+void Net::Impl::initBackend(const std::vector<LayerPin>& blobsToKeep_)
+{
+    CV_TRACE_FUNCTION();
+    if (preferableBackend == DNN_BACKEND_OPENCV)
+    {
+        CV_Assert(preferableTarget == DNN_TARGET_CPU || IS_DNN_OPENCL_TARGET(preferableTarget));
```

Comment:
We could add initOpenCVBackend() as well. This would help get rid of copies in situations where blob preparation is destructive, so we have to keep a copy for some other backend to act on.